File: 백준/Gold/1516. 게임 개발/게임 개발.cc
#include <bits/stdc++.h>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int N;
    cin >> N;

    vector<int> time(N + 1);
    vector<int> indegree(N + 1);      // prev 개수
    vector<vector<int>> graph(N + 1); // prev: [post]
    vector<int> dp(N + 1);
    for (int i = 1; i <= N; i++) {
        cin >> time[i];
        int j;
        while (cin >> j) {
            if (j == -1)
                break;
            indegree[i]++;
            graph[j].push_back(i);
        }
    }

    queue<int> q;
    for (int i = 1; i <= N; i++) {
        if (indegree[i] == 0) {
            dp[i] = time[i];
            q.push(i);
        }
    }

    while (!q.empty()) {
        int curr = q.front();
        q.pop();

        for (int nxt : graph[curr]) {
            dp[nxt] = max(dp[nxt], dp[curr] + time[nxt]);
            indegree[nxt]--;
            if (indegree[nxt] == 0)
                q.push(nxt);
        }
    }

    for (int i = 1; i <= N; i++) {
        cout << dp[i] << '\n';
    }

    return 0;
}

File: 백준/Gold/1516. 게임 개발/게임 개발.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input())
time = [0] * (N + 1)  # 1번 건물을 짓는데 걸리는 시간
graph = {i: [] for i in range(1, N + 1)}  # prev: [post]
indegree = [0] * (N + 1)  # i번 건물의 선행 건물 개수
dp = [0] * (N + 1)  # i번 건물을 짓기 위한 최소 시간
for i in range(1, N + 1):
    t, *prevs = map(int, input().split())
    prevs.pop()

    time[i] = t
    indegree[i] += len(prevs)
    for prev in prevs:
        graph[prev].append(i)

q = deque()
# 선행 건물이 없는 건물 추가
for i in range(1, N + 1):
    if indegree[i] == 0:
        dp[i] = time[i]
        q.append(i)

while q:
    curr = q.popleft()

    for nxt in graph[curr]:
        dp[nxt] = max(dp[nxt], dp[curr] + time[nxt])
        indegree[nxt] -= 1

        if indegree[nxt] == 0:
            q.append(nxt)

for i in range(1, N + 1):
    print(dp[i])
